File: app/src/main/java/com/example/bloodbank/DonorRegistrationActivity.java
package com.example.bloodbank;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;

import android.app.DatePickerDialog;
import android.app.ProgressDialog;
import android.content.Intent;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.text.TextUtils;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.OnFailureListener;
import com.google.android.gms.tasks.OnSuccessListener;
import com.google.android.gms.tasks.Task;
import com.google.android.material.textfield.MaterialAutoCompleteTextView;
import com.google.android.material.textfield.TextInputEditText;
import com.google.firebase.auth.AuthResult;
import com.google.firebase.auth.FirebaseAuth;
import com.google.firebase.auth.FirebaseUser;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.storage.FirebaseStorage;
import com.google.firebase.storage.StorageReference;
import com.google.firebase.storage.UploadTask;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;
import java.util.Locale;

import de.hdodenhof.circleimageview.CircleImageView;

public class DonorRegistrationActivity extends AppCompatActivity {

    private TextView backButton;

    private CircleImageView profile_image;
    private TextInputEditText registeredFullName, registeredIdNumber, registeredPhoneNumber,
            registeredOccupation, registeredEmail, registeredPassword, registeredLastDonation,
            registeredAddress, registeredBirthdate;
    private MaterialAutoCompleteTextView bloodGroupsSpinner;
    private MaterialAutoCompleteTextView gendersSpinner;

    private Button registerButton;

    private Uri resultUri;

    private ProgressDialog loader;

    private FirebaseAuth mAuth;
    private DatabaseReference userDatabaseRef;

    private Calendar calendar;
    private SimpleDateFormat dateFormat;

    private static final String APP_DIRECTORY = "BloodBank";
    private static final String IMAGE_DIRECTORY = "ProfileImages";
    private File currentPhotoFile;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_donor_registration);

        backButton = findViewById(R.id.backButton);
        backButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(DonorRegistrationActivity.this, LoginActivity.class);
                startActivity(intent);
            }
        });

        profile_image = findViewById(R.id.profile_image);
        registeredFullName = findViewById(R.id.registeredFullName);
        registeredIdNumber = findViewById(R.id.registeredIdNumber);
        registeredPhoneNumber = findViewById(R.id.registeredPhoneNumber);
        registeredOccupation = findViewById(R.id.registeredOccupation);
        registeredEmail = findViewById(R.id.registeredEmail);
        registeredPassword = findViewById(R.id.registeredPassword);
        registeredLastDonation = findViewById(R.id.registeredLastDonation);
        registeredAddress = findViewById(R.id.registeredAddress);
        registeredBirthdate = findViewById(R.id.registeredDate);
        gendersSpinner = findViewById(R.id.gendersSpinner);
        bloodGroupsSpinner = findViewById(R.id.bloodGroupsSpinner);
        registerButton = findViewById(R.id.registerButton);
        loader = new ProgressDialog(this);

        mAuth = FirebaseAuth.getInstance();
        userDatabaseRef = FirebaseDatabase.getInstance().getReference();

        // Initialize gender spinner
        String[] genderItems = new String[]{"Select your gender", "Male", "Female", "Other"};
        ArrayAdapter<String> genderAdapter = new ArrayAdapter<>(this, android.R.layout.simple_dropdown_item_1line, genderItems);
        gendersSpinner.setAdapter(genderAdapter);

        // Initialize blood group spinner
        String[] bloodGroupItems = new String[]{"Select your blood group", "A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-"};
        ArrayAdapter<String> bloodGroupAdapter = new ArrayAdapter<>(this, android.R.layout.simple_dropdown_item_1line, bloodGroupItems);
        bloodGroupsSpinner.setAdapter(bloodGroupAdapter);

        // Initialize date picker components
        calendar = Calendar.getInstance();
        dateFormat = new SimpleDateFormat("dd/MM/yyyy", Locale.getDefault());

        // Set up date picker for birthdate
        registeredBirthdate.setOnClickListener(v -> showDatePicker(registeredBirthdate));
        registeredBirthdate.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus) {
                showDatePicker(registeredBirthdate);
            }
        });

        // Set up date picker for last donation
        registeredLastDonation.setOnClickListener(v -> showDatePicker(registeredLastDonation));
        registeredLastDonation.setOnFocusChangeListener((v, hasFocus) -> {
            if (hasFocus) {
                showDatePicker(registeredLastDonation);
            }
        });

        profile_image.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(Intent.ACTION_PICK);
                intent.setType("image/*");
                startActivityForResult(intent, 1);
            }
        });

        registerButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                final String email = registeredEmail.getText().toString().trim();
                final String password = registeredPassword.getText().toString().trim();
                final String fullName = registeredFullName.getText().toString().trim();
                final String idNumber = registeredIdNumber.getText().toString().trim();
                final String phoneNumber = registeredPhoneNumber.getText().toString().trim();
                final String occupation = registeredOccupation.getText().toString().trim();
                final String lastDonation = registeredLastDonation.getText().toString().trim();
                final String address = registeredAddress.getText().toString().trim();
                final String birthdate = registeredBirthdate.getText().toString().trim();
                final String gender = gendersSpinner.getText().toString();
                final String bloodGroup = bloodGroupsSpinner.getText().toString();

                if (TextUtils.isEmpty(email)) {
                    registeredEmail.setError("Email is required!");
                    return;
                }
                if (TextUtils.isEmpty(password)) {
                    registeredPassword.setError("Password is required!");
                    return;
                }
                if (TextUtils.isEmpty(fullName)) {
                    registeredFullName.setError("Full Name is required!");
                    return;
                }
                if (TextUtils.isEmpty(idNumber)) {
                    registeredIdNumber.setError("National ID Number is required!");
                    return;
                }
                if (TextUtils.isEmpty(phoneNumber)) {
                    registeredPhoneNumber.setError("Phone Number is required!");
                    return;
                }
                if (TextUtils.isEmpty(occupation)) {
                    registeredOccupation.setError("Occupation is required!");
                    return;
                }
                if (TextUtils.isEmpty(address)) {
                    registeredAddress.setError("Address is required!");
                    return;
                }
                if (TextUtils.isEmpty(birthdate)) {
                    registeredBirthdate.setError("Birthdate is required!");
                    return;
                }
                if (TextUtils.isEmpty(lastDonation)) {
                    registeredLastDonation.setError("Last donation date is required!");
                    return;
                }
                if (gender.equals("Select your gender")) {
                    Toast.makeText(DonorRegistrationActivity.this, "Please select your gender", Toast.LENGTH_SHORT).show();
                    return;
                }
                if (bloodGroup.equals("Select your blood group")) {
                    Toast.makeText(DonorRegistrationActivity.this, "Please select your blood group", Toast.LENGTH_SHORT).show();
                    return;
                }

                loader.setMessage("Registering you...");
                loader.setCanceledOnTouchOutside(false);
                loader.show();

                mAuth.createUserWithEmailAndPassword(email, password).addOnCompleteListener(new OnCompleteListener<AuthResult>() {
                    @Override
                    public void onComplete(@NonNull Task<AuthResult> task) {
                        if (task.isSuccessful()) {
                            FirebaseUser user = mAuth.getCurrentUser();
                            if (user != null) {
                                // Send email verification
                                user.sendEmailVerification().addOnSuccessListener(new OnSuccessListener<Void>() {
                                    @Override
                                    public void onSuccess(Void unused) {
                                        Toast.makeText(DonorRegistrationActivity.this, "Verification Email has been sent.", Toast.LENGTH_SHORT).show();
                                    }
                                }).addOnFailureListener(new OnFailureListener() {
                                    @Override
                                    public void onFailure(@NonNull Exception e) {
                                        Toast.makeText(DonorRegistrationActivity.this, "Email verification failed: " + e.getMessage(), Toast.LENGTH_SHORT).show();
                                    }
                                });

                                // Save user data to database
                                String currentUserId = user.getUid();
                                userDatabaseRef = FirebaseDatabase.getInstance().getReference()
                                        .child("users").child(currentUserId);
                                HashMap userInfo = new HashMap();
                                userInfo.put("id", currentUserId);
                                userInfo.put("name", fullName);
                                userInfo.put("email", email);
                                userInfo.put("idnumber", idNumber);
                                userInfo.put("phoneNumber", phoneNumber);
                                userInfo.put("password", password);
                                userInfo.put("address", address);
                                userInfo.put("birthdate", birthdate);
                                userInfo.put("gender", gender);
                                userInfo.put("bloodGroup", bloodGroup);
                                userInfo.put("type", "donor");
                                userInfo.put("search", "donor"+bloodGroup);
                                userInfo.put("occupation", occupation);
                                userInfo.put("lastDonationDate", lastDonation);

                                // If we have a profile image, save its path
                                if (currentPhotoFile != null && currentPhotoFile.exists()) {
                                    userInfo.put("profileImagePath", currentPhotoFile.getAbsolutePath());
                                }

                                userDatabaseRef.updateChildren(userInfo).addOnCompleteListener(new OnCompleteListener() {
                                    @Override
                                    public void onComplete(@NonNull Task task) {
                                        if (task.isSuccessful()) {
                                            Toast.makeText(DonorRegistrationActivity.this, "Registration successful!", Toast.LENGTH_SHORT).show();
                                            Intent intent = new Intent(DonorRegistrationActivity.this, VerifyEmailActivity.class);
                                            startActivity(intent);
                                            finish();
                                        } else {
                                            Toast.makeText(DonorRegistrationActivity.this, "Database error: " + task.getException().getMessage(), Toast.LENGTH_SHORT).show();
                                        }
                                        loader.dismiss();
                                    }
                                });
                            }
                        } else {
                            String errorMessage = task.getException() != null ? task.getException().getMessage() : "Unknown error";
                            Toast.makeText(DonorRegistrationActivity.this, "Registration failed: " + errorMessage, Toast.LENGTH_SHORT).show();
                            loader.dismiss();
                        }
                    }
                });
            }
        });
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 1 && resultCode == RESULT_OK && data != null) {
            resultUri = data.getData();
            profile_image.setImageURI(resultUri);
            
            try {
                // Create directory if it doesn't exist
                File storageDir = new File(getFilesDir(), APP_DIRECTORY + File.separator + IMAGE_DIRECTORY);
                if (!storageDir.exists()) {
                    storageDir.mkdirs();
                }

                // Create a unique filename
                String timeStamp = new SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(new Date());
                String imageFileName = "JPEG_" + timeStamp + "_";
                currentPhotoFile = File.createTempFile(
                    imageFileName,
                    ".jpg",
                    storageDir
                );

                // Save the image to internal storage
                Bitmap bitmap = MediaStore.Images.Media.getBitmap(getContentResolver(), resultUri);
                FileOutputStream out = new FileOutputStream(currentPhotoFile);
                bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out);
                out.close();
            } catch (IOException e) {
                e.printStackTrace();
                Toast.makeText(DonorRegistrationActivity.this, "Error saving profile image", Toast.LENGTH_SHORT).show();
            }
        }
    }

    private void showDatePicker(TextInputEditText targetField) {
        DatePickerDialog.OnDateSetListener dateSetListener = (view, year, month, dayOfMonth) -> {
            calendar.set(Calendar.YEAR, year);
            calendar.set(Calendar.MONTH, month);
            calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth);
            updateDateField(targetField);
        };

        new DatePickerDialog(
            this,
            dateSetListener,
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        ).show();
    }

    private void updateDateField(TextInputEditText targetField) {
        targetField.setText(dateFormat.format(calendar.getTime()));
    }

    private boolean isValidDate(String dateStr, boolean isBirthDate) {
        try {
            Date date = dateFormat.parse(dateStr);
            Calendar inputDate = Calendar.getInstance();
            inputDate.setTime(date);
            
            // Get current date
            Calendar currentDate = Calendar.getInstance();
            
            if (isBirthDate) {
                // Check if birth date is in the future
                if (inputDate.after(currentDate)) {
                    registeredBirthdate.setError("Birth date cannot be in the future");
                    return false;
                }
                
                // Check if age is at least 18 years
                Calendar minAgeDate = Calendar.getInstance();
                minAgeDate.add(Calendar.YEAR, -18);
                if (inputDate.after(minAgeDate)) {
                    registeredBirthdate.setError("You must be at least 18 years old");
                    return false;
                }
            } else {
                // Check if last donation date is in the future
                if (inputDate.after(currentDate)) {
                    registeredLastDonation.setError("Last donation date cannot be in the future");
                    return false;
                }
                
                // Check if date is within last 3 months (blood donation eligibility)
                Calendar threeMonthsAgo = Calendar.getInstance();
                threeMonthsAgo.add(Calendar.MONTH, -3);
                if (inputDate.after(threeMonthsAgo)) {
                    registeredLastDonation.setError("You must wait 3 months between donations");
                    return false;
                }
            }
            
            return true;
        } catch (Exception e) {
            if (isBirthDate) {
                registeredBirthdate.setError("Invalid date format");
            } else {
                registeredLastDonation.setError("Invalid date format");
            }
            return false;
        }
    }
}

File: app/src/main/java/com/example/bloodbank/DonorRegistrationActivity.kt
package com.example.bloodbank

import android.app.DatePickerDialog
import android.content.Intent
import android.graphics.Bitmap
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.example.bloodbank.Model.User
import com.google.android.material.textfield.MaterialAutoCompleteTextView
import com.google.android.material.textfield.TextInputEditText
import dagger.hilt.android.AndroidEntryPoint
import de.hdodenhof.circleimageview.CircleImageView
import kotlinx.coroutines.launch
import java.io.File
import java.io.FileOutputStream
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale

/**
 * DonorRegistrationActivity - Donor registration with comprehensive validation.
 * Follows MVVM architecture with AuthViewModel.
 */
@AndroidEntryPoint
class DonorRegistrationActivity : AppCompatActivity() {
    
    private val viewModel: AuthViewModel by viewModels()
    
    // Views
    private lateinit var backButton: TextView
    private lateinit var profileImage: CircleImageView
    private lateinit var fullNameInput: TextInputEditText
    private lateinit var idNumberInput: TextInputEditText
    private lateinit var phoneInput: TextInputEditText
    private lateinit var occupationInput: TextInputEditText
    private lateinit var emailInput: TextInputEditText
    private lateinit var passwordInput: TextInputEditText
    private lateinit var lastDonationInput: TextInputEditText
    private lateinit var addressInput: TextInputEditText
    private lateinit var birthdateInput: TextInputEditText
    private lateinit var bloodGroupSpinner: MaterialAutoCompleteTextView
    private lateinit var genderSpinner: MaterialAutoCompleteTextView
    private lateinit var registerButton: Button
    
    // State
    private var currentPhotoFile: File? = null
    private val calendar = Calendar.getInstance()
    private val dateFormat = SimpleDateFormat("dd/MM/yyyy", Locale.getDefault())
    
    // Modern image picker
    private val imagePickerLauncher = registerForActivityResult(
        ActivityResultContracts.GetContent()
    ) { uri: Uri? ->
        uri?.let { handleImageSelection(it) }
    }
    
    companion object {
        private const val APP_DIRECTORY = "BloodBank"
        private const val IMAGE_DIRECTORY = "ProfileImages"
    }
    
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_donor_registration)
        
        initializeViews()
        setupSpinners()
        setupDatePickers()
        setupClickListeners()
        observeViewModel()
    }
    
    private fun initializeViews() {
        backButton = findViewById(R.id.backButton)
        profileImage = findViewById(R.id.profile_image)
        fullNameInput = findViewById(R.id.registeredFullName)
        idNumberInput = findViewById(R.id.registeredIdNumber)
        phoneInput = findViewById(R.id.registeredPhoneNumber)
        occupationInput = findViewById(R.id.registeredOccupation)
        emailInput = findViewById(R.id.registeredEmail)
        passwordInput = findViewById(R.id.registeredPassword)
        lastDonationInput = findViewById(R.id.registeredLastDonation)
        addressInput = findViewById(R.id.registeredAddress)
        birthdateInput = findViewById(R.id.registeredDate)
        genderSpinner = findViewById(R.id.gendersSpinner)
        bloodGroupSpinner = findViewById(R.id.bloodGroupsSpinner)
        registerButton = findViewById(R.id.registerButton)
    }
    
    private fun setupSpinners() {
        // Gender spinner
        val genderItems = arrayOf("Select your gender", "Male", "Female", "Other")
        val genderAdapter = ArrayAdapter(this, android.R.layout.simple_dropdown_item_1line, genderItems)
        genderSpinner.setAdapter(genderAdapter)
        
        // Blood group spinner
        val bloodGroupItems = arrayOf("Select your blood group", "A+", "A-", "B+", "B-", "AB+", "AB-", "O+", "O-")
        val bloodGroupAdapter = ArrayAdapter(this, android.R.layout.simple_dropdown_item_1line, bloodGroupItems)
        bloodGroupSpinner.setAdapter(bloodGroupAdapter)
    }
    
    private fun setupDatePickers() {
        birthdateInput.setOnClickListener { showDatePicker(birthdateInput, true) }
        birthdateInput.setOnFocusChangeListener { _, hasFocus ->
            if (hasFocus) showDatePicker(birthdateInput, true)
        }
        
        lastDonationInput.setOnClickListener { showDatePicker(lastDonationInput, false) }
        lastDonationInput.setOnFocusChangeListener { _, hasFocus ->
            if (hasFocus) showDatePicker(lastDonationInput, false)
        }
    }
    
    private fun setupClickListeners() {
        backButton.setOnClickListener {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }
        
        profileImage.setOnClickListener {
            imagePickerLauncher.launch("image/*")
        }
        
        registerButton.setOnClickListener {
            if (validateInputs()) {
                registerDonor()
            }
        }
    }
    
    private fun validateInputs(): Boolean {
        val email = emailInput.text.toString().trim()
        val password = passwordInput.text.toString().trim()
        val fullName = fullNameInput.text.toString().trim()
        val idNumber = idNumberInput.text.toString().trim()
        val phone = phoneInput.text.toString().trim()
        val occupation = occupationInput.text.toString().trim()
        val address = addressInput.text.toString().trim()
        val birthdate = birthdateInput.text.toString().trim()
        val lastDonation = lastDonationInput.text.toString().trim()
        val gender = genderSpinner.text.toString()
        val bloodGroup = bloodGroupSpinner.text.toString()
        
        return when {
            email.isEmpty() -> {
                emailInput.error = "Email is required"
                false
            }
            password.isEmpty() -> {
                passwordInput.error = "Password is required"
                false
            }
            password.length < 6 -> {
                passwordInput.error = "Password must be at least 6 characters"
                false
            }
            fullName.isEmpty() -> {
                fullNameInput.error = "Full name is required"
                false
            }
            idNumber.isEmpty() -> {
                idNumberInput.error = "ID number is required"
                false
            }
            phone.isEmpty() -> {
                phoneInput.error = "Phone number is required"
                false
            }
            occupation.isEmpty() -> {
                occupationInput.error = "Occupation is required"
                false
            }
            address.isEmpty() -> {
                addressInput.error = "Address is required"
                false
            }
            birthdate.isEmpty() -> {
                birthdateInput.error = "Birthdate is required"
                false
            }
            lastDonation.isEmpty() -> {
                lastDonationInput.error = "Last donation date is required"
                false
            }
            gender == "Select your gender" -> {
                Toast.makeText(this, "Please select your gender", Toast.LENGTH_SHORT).show()
                false
            }
            bloodGroup == "Select your blood group" -> {
                Toast.makeText(this, "Please select your blood group", Toast.LENGTH_SHORT).show()
                false
            }
            else -> true
        }
    }
    
    private fun registerDonor() {
        val user = User(
            id = "",
            name = fullNameInput.text.toString().trim(),
            email = emailInput.text.toString().trim(),
            idnumber = idNumberInput.text.toString().trim(),
            phoneNumber = phoneInput.text.toString().trim(),
            address = addressInput.text.toString().trim(),
            birthdate = birthdateInput.text.toString().trim(),
            gender = genderSpinner.text.toString(),
            bloodGroup = bloodGroupSpinner.text.toString(),
            type = "donor",
            search = "donor${bloodGroupSpinner.text}",
            occupation = occupationInput.text.toString().trim(),
            lastDonationDate = lastDonationInput.text.toString().trim(),
            profileImagePath = currentPhotoFile?.absolutePath ?: ""
        )
        
        val password = passwordInput.text.toString().trim()
        viewModel.register(user, password)
    }
    
    private fun observeViewModel() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    viewModel.authState.collect { state ->
                        when (state) {
                            is AuthViewModel.AuthState.Authenticated -> {
                                Toast.makeText(
                                    this@DonorRegistrationActivity,
                                    "Registration successful! Please verify your email.",
                                    Toast.LENGTH_LONG
                                ).show()
                                startActivity(Intent(this@DonorRegistrationActivity, VerifyEmailActivity::class.java))
                                finish()
                            }
                            else -> {}
                        }
                    }
                }
                
                launch {
                    viewModel.isLoading.collect { isLoading ->
                        registerButton.isEnabled = !isLoading
                        registerButton.text = if (isLoading) "Registering..." else "Register"
                    }
                }
                
                launch {
                    viewModel.error.collect { error ->
                        error?.let {
                            Toast.makeText(this@DonorRegistrationActivity, it, Toast.LENGTH_SHORT).show()
                            viewModel.clearError()
                        }
                    }
                }
            }
        }
    }
    
    private fun showDatePicker(targetField: TextInputEditText, isBirthdate: Boolean) {
        DatePickerDialog(
            this,
            { _, year, month, dayOfMonth ->
                calendar.set(Calendar.YEAR, year)
                calendar.set(Calendar.MONTH, month)
                calendar.set(Calendar.DAY_OF_MONTH, dayOfMonth)
                targetField.setText(dateFormat.format(calendar.time))
            },
            calendar.get(Calendar.YEAR),
            calendar.get(Calendar.MONTH),
            calendar.get(Calendar.DAY_OF_MONTH)
        ).show()
    }
    
    private fun handleImageSelection(uri: Uri) {
        try {
            profileImage.setImageURI(uri)
            
            // Create directory
            val storageDir = File(filesDir, "$APP_DIRECTORY${File.separator}$IMAGE_DIRECTORY").apply {
                if (!exists()) mkdirs()
            }
            
            // Create unique filename
            val timeStamp = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.getDefault()).format(Date())
            currentPhotoFile = File.createTempFile("JPEG_${timeStamp}_", ".jpg", storageDir)
            
            // Save image
            val bitmap = MediaStore.Images.Media.getBitmap(contentResolver, uri)
            FileOutputStream(currentPhotoFile).use { out ->
                bitmap.compress(Bitmap.CompressFormat.JPEG, 90, out)
            }
        } catch (e: Exception) {
            Toast.makeText(this, "Error saving profile image", Toast.LENGTH_SHORT).show()
        }
    }
}
